llvm/include/llvm/ADT/PagedVector.h

```diff
@@ -0,0 +1,301 @@
+    Size = NewSize;
+  }
+
+  // Return true if the vector is empty
+  bool empty() const { return Size == 0; }
+
+  /// Clear the vector, i.e. clear the allocated pages, the whole page
+  /// lookup index and reset the size.
+  void clear() {
+    Size = 0;
+    // If we own the allocator, simply reset it, otherwise we
+    // deallocate the pages one by one.
+    if (Allocator & 0x1) {
+      getAllocator()->Reset();
+    } else {
+      for (auto *Page : PageToDataIdx) {
+        getAllocator()->Deallocate(Page);
+      }
+    }
+    PageToDataIdx.clear();
+  }
+
+  // Iterator on all the elements of the vector
+  // which have actually being constructed.
+  class MaterialisedIterator {
+    PagedVector const *PV;
+    size_t ElementIdx;
+
+  public:
+    using iterator_category = std::forward_iterator_tag;
+    using value_type = T;
+    using difference_type = std::ptrdiff_t;
+    using pointer = T *;
+    using reference = T &;
+
+    MaterialisedIterator(PagedVector const *PV, size_t ElementIdx)
+        : PV(PV), ElementIdx(ElementIdx) {}
+
+    // When incrementing the iterator, we skip the elements which have not
+    // been materialised yet.
+    MaterialisedIterator &operator++() {
+      while (ElementIdx < PV->Size) {
+        ++ElementIdx;
+        if (PV->PageToDataIdx[ElementIdx / PAGE_SIZE] != invalidPage()) {
+          return *this;
+        }
+      }
+      return *this;
+    }
+    // Post increment operator.
+    MaterialisedIterator operator++(int) {
+      auto Copy = *this;
+      ++*this;
+      return Copy;
+    }
+
+    std::ptrdiff_t operator-(MaterialisedIterator const &Other) const {
+      assert(PV == Other.PV);
+      // If they are on the same table we can just subtract the indices.
+      // Otherwise we have to iterate over the pages to find the difference.
+      // If a page is invalid, we skip it.
+      if (PV == Other.PV) {
+        return ElementIdx - Other.ElementIdx;
+      }
+
+      auto ElementMin = std::min(ElementIdx, Other.ElementIdx);
+      auto ElementMax = std::max(ElementIdx, Other.ElementIdx);
+      auto PageMin = ElementMin / PAGE_SIZE;
+      auto PageMax = ElementMax / PAGE_SIZE;
+
+      auto Count = 0ULL;
+      for (auto PageIdx = PageMin; PageIdx < PageMax; ++PageIdx) {
+        if (PV->PageToDataIdx[PageIdx] == invalidPage()) {
+          continue;
+        }
+        Count += PAGE_SIZE;
+      }
+      Count += ElementMax % PAGE_SIZE;
+      Count += PAGE_SIZE - ElementMin % PAGE_SIZE;
+
+      return Count;
+    }
+
+    // When dereferencing the iterator, we materialise the page if needed.
+    T const &operator*() const {
+      assert(ElementIdx < PV->Size);
+      assert(PV->PageToDataIdx[ElementIdx / PAGE_SIZE] != invalidPage());
+      return *((ElementIdx % PAGE_SIZE) +
+               PV->PageToDataIdx[ElementIdx / PAGE_SIZE]);
+    }
+
+    // Equality operator.
+    bool operator==(MaterialisedIterator const &Other) const {
```

Comment:
Prefer non-member (can be a friend, which allows you to keep the definition here/inline) so the LHS and RHS conversions are equally viable. Same feedback for other operator overloads in this patch. (when they can be non-member, they should be - some must be members, which don't apply to this feedback)

Reply:
I think I have fixed what should have been fixed.